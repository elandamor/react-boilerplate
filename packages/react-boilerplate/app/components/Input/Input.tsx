import React, { SFC } from 'react';
import classNames from 'classnames';
// Styles
import Wrapper from './styles';

/**
 * @render react
 * @name Input component
 * @description Input component.
 * @example
 * <Input
 *  id="text"
 *  label="Text"
 *  type="text"
 * />
 */

interface IProps {
  className?: string;
  id: string;
  label: string;
  sronly?: boolean;
  type?: string;
}

const Input: SFC<IProps> = ({ className, id, label, sronly, ...rest }) => (
  <Wrapper className={classNames('c-input-wrapper', className)}>
    <label htmlFor={id}>
      <span
        className={classNames('a-label', {
          'sr-only': sronly,
        })}
      >
        { label }
      </span>
      <input id={id} className="a-input" {...rest} />
    </label>
  </Wrapper>
);

Input.defaultProps = {
  sronly: false,
  type: 'text',
};

export default Input;
