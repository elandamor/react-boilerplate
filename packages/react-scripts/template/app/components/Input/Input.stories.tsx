import React from 'react';
import { storiesOf } from '@storybook/react';

import Input from './index';

storiesOf('Input', module)
.add('text', () => (
  <Input
    id="name"
    label="Name"
    name="name"
    onChange={() => null}
    placeholder="Name"
    type="text"
    helperText="Helper text"
    sronly
  />
))
.add('checkbox', () => (
  <Input
    id="checkbox"
    label="Checkbox"
    name="checkbox"
    onChange={() => null}
    type="checkbox"
  />
))
.add('radio', () => (
  <Input
    id="radio"
    label="Radio"
    name="radio"
    onChange={() => null}
    type="radio"
  />
))
.add('toggle', () => (
  <Input
    id="toggle"
    label="Toggle"
    name="toggle"
    onChange={() => null}
    type="checkbox"
    as="toggle"
  />
))
