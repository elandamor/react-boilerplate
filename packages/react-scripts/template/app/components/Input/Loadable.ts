/**
 *
 * Asynchronously loads the component for Input
 *
 */

import React from 'react';

export default React.lazy(() => import('./index'));
