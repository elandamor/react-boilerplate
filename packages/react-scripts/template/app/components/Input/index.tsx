import React, { SFC } from 'react';
import classNames from 'classnames';
// Styles
import Wrapper from './styles';

/**
 * @render react
 * @name Input component
 * @description Input component.
 * @example
 * <Input
 *  id="text"
 *  label="Text"
 *  name="text"
 *  type="text"
 * />
 */

interface IProps {
  className?: string;
  checked?: boolean;
  id: string;
  label: string;
  name: string;
  onChange: (event: any) => void;
  placeholder?: string;
  readonly?: boolean;
  rows?: number;
  sronly?: boolean;
  type:
    | 'text'
    | 'email'
    | 'file'
    | 'date'
    | 'password'
    | 'textarea'
    | 'checkbox'
    | 'radio'
    | 'number';
  value?: any;
};

const Input: SFC<IProps> = ({
  checked,
  className,
  id,
  label,
  sronly,
  ...rest
}) => (
  <Wrapper
    className={classNames('c-input__wrapper', className, { '-active': checked })}
  >
    <label htmlFor={id}>
      <span
        className={classNames('a-label', {
          'sr-only': sronly,
        })}
      >
        {label}
      </span>
      {rest.type === 'textarea' ? (
        <textarea id={id} className="a-textarea" {...rest} />
      ) : (
        <input id={id} className="a-input" {...rest} />
      )}
    </label>
  </Wrapper>
);

export default Input;
