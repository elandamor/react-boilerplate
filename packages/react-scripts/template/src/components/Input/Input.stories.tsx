import React from 'react';
import { storiesOf } from '@storybook/react';

import Input from './index';

storiesOf('Input', module)
.add('text', () => (
  <Input
    id="name"
    label="Name"
    name="name"
    onChange={() => null}
    type="text"
  />
))
.add('checkbox', () => (
  <Input
    id="checkbox"
    label="Checkbox"
    name="checkbox"
    onChange={() => null}
    type="checkbox"
  />
))
.add('radio', () => (
  <Input
    id="radio"
    label="Radio"
    name="radio"
    onChange={() => null}
    type="radio"
  />
))
.add('toggle', () => (
  <Input
    id="toggle"
    label="Toggle"
    name="toggle"
    onChange={() => null}
    type="checkbox"
    as="toggle"
  />
))
.add('dropzone', () => (
  <Input
    id="file"
    label="File"
    name="file"
    onChange={() => null}
    type="file"
  />
))
.add('select', () => (
  <Input
    id="select"
    label="Select"
    name="select"
    onChange={() => null}
    type="select"
  />
))
