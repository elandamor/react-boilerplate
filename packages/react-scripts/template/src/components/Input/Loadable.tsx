/**
 *
 * Asynchronously loads the component for Input
 *
 */

import React from 'react';
import { Loadable } from '../../utils';
import LoadingBar from '../LoadingBar';

export default Loadable(() => import('./index'), {
  fallback: <LoadingBar />,
});
