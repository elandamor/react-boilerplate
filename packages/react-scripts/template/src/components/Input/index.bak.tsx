import React, { FC, Suspense } from 'react';
import classNames from 'classnames';
// Styles
import Wrapper, { HelperText } from './styles';
import Checkbox from '../Checkbox/Loadable';
import Dropzone from '../Dropzone/Loadable';
import Label from '../Label';
import LoadingBar from '../LoadingBar';
import Select from '../Select/Loadable';
import Spacer from '../Spacer';
import theme from '../../theme';

export interface IInputProps {
  as?: string;
  className?: string;
  checked?: boolean;
  helperText?: string;
  helperTextPosition?: 'bottom' | 'top';
  id: string;
  label: string;
  name: string;
  onChange?: (event: React.ChangeEvent<HTMLElement>) => void;
  placeholder?: string;
  readonly?: boolean;
  rows?: number;
  sronly?: boolean;
  type:
    | 'text'
    | 'email'
    | 'file'
    | 'date'
    | 'password'
    | 'textarea'
    | 'checkbox'
    | 'radio'
    | 'number'
    | 'select';
  value?: any;
}

/**
 * @render react
 * @name Input component
 * @description Input component.
 * @example
 * <Input
 *  id="text"
 *  label="Text"
 *  name="text"
 *  type="text"
 * />
 */

const Input: FC<IInputProps> = ({
  checked,
  className,
  helperText,
  id,
  label,
  onChange: handleChange,
  sronly,
  ...rest
}) => {
  const renderLabel = () => <Label sronly={sronly}>{label}</Label>;

  const renderInput = () => {
    switch (rest.type) {
      case 'checkbox':
        return <Checkbox onChange={handleChange} {...rest} />
      case 'file':
        return <Dropzone onChange={handleChange} {...rest} />;
      case 'select':
        return <Select onChange={handleChange} {...rest} />;
      case 'textarea':
        return <textarea id={id} className="a-textarea" aria-label={label} onChange={handleChange} {...rest} />;
      default:
        return <input id={id} className="a-input" aria-label={label} onChange={handleChange} {...rest} />;
    }
  };

  return (
    <Wrapper
      className={classNames('c-input__wrapper', className, {
        '-active': checked,
      })}
    >
      <label htmlFor={id}>
        {
          (rest.type !== 'checkbox' && rest.type !== 'radio')
          && (
            <React.Fragment>
              {renderLabel()}
              <Spacer spacing={theme.space[1]} />
            </React.Fragment>
          )
        }
        {
          rest.helperTextPosition === 'top' && helperText
          && <HelperText>{helperText}</HelperText>
        }
        <Suspense fallback={<LoadingBar />}>{ renderInput() }</Suspense>
        {
          (rest.type === 'checkbox' || rest.type === 'radio')
          && (
            <React.Fragment>
              <span className={`a-${rest.type}`} />
              {renderLabel()}
            </React.Fragment>
          )
        }
        {
          rest.helperTextPosition === 'bottom' && helperText
          && <HelperText>{helperText}</HelperText>
        }
      </label>
    </Wrapper>
  );
};

Input.defaultProps = {
  as: undefined,
  className: '',
  checked: false,
  helperText: '',
  helperTextPosition: 'bottom',
  onChange: () => null,
  placeholder: '',
  readonly: false,
  rows: 5,
  sronly: false,
  value: ''
};

export default Input;
