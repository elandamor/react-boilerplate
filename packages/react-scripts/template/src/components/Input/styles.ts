import styled, { css } from 'styled-components';
import { IInputProps } from './index';
import theme from '../../theme';

const Wrapper = styled.div<IInputProps>`
  margin-bottom: ${theme.space[2]}px;
  position: relative;

  .a-checkbox,
  .a-radio {
    border: ${theme.borders[2]} #969696;
    border-radius: 50%;
    cursor: pointer;
    display: inline-block;
    height: 1.25rem;
    margin-right: ${theme.space[1]}px;
    position: relative;
    vertical-align: middle;
    width: 1.25rem;
  }

  .a-checkbox {
    ${({ as }) => as === 'toggle' && css`
      border: none;
      height: 1.25rem;
      width: 2rem;

      &::after, &::before {
        content: '';
        top: 0;
        left: 0;
        position: absolute;
      }

      &::before {
        width: 100%;
        height: 100%;
        border-radius: 0.75rem;
        background: #dbdbdb;
      }

      &::after {
        width: 1rem;
        height: 1rem;
        margin: 0.15rem;
        border-radius: 50%;
        background: #ffffff;
      }
    `}
  }

  input, textarea {
    background: transparent;
    border: ${theme.borders[1]} #aaaaaa;
    border-radius: ${theme.space[1] / 2}px;
    font-size: ${theme.fontSizes[2]}px;
    padding: ${theme.space[2]}px ${theme.space[1] + theme.space[1] / 2}px;
    position: relative;
    width: 100%;
    z-index: 1;

    ${({ type }) => (type === 'checkbox' || type === 'radio')
    && css`
      top: 0;
      left: 0;
      width: 0;
      height: 0;
      opacity: 0;
      filter: alpha(opacity=0);
      position: absolute;
      visibility: hidden;
    `}

    &:focus,
    &:hover {
      border: ${theme.borders[1]} ${theme.colors.primary};
      outline: none;
    }

    &:focus {
      font-size: ${theme.fontSizes[4]}px;
    }

    &:hover ~ .a-checkbox {
      border-color: #4a4a4a;
    }

    &:hover ~ .a-radio {
      border: ${theme.borders[2]} #4a4a4a;
    }

    &:checked ~ .a-checkbox {
      border-color: ${theme.colors.primary};
      background: ${theme.colors.primary} url(data:image/svg+xml;base64,PHN2ZyB4bWxucz0iaHR0cDovL3d3dy53My5vcmcvMjAwMC9zdmciIHdpZHRoPSIxNiIgaGVpZ2h0PSIxNiIgdmlld0JveD0iMCAwIDI0IDI0IiBmaWxsPSJub25lIiBzdHJva2U9IiNmZmZmZmYiIHN0cm9rZS13aWR0aD0iMyIgc3Ryb2tlLWxpbmVjYXA9InJvdW5kIiBzdHJva2UtbGluZWpvaW49InJvdW5kIj48cG9seWxpbmUgcG9pbnRzPSIyMCA2IDkgMTcgNCAxMiI+PC9wb2x5bGluZT48L3N2Zz4=) center no-repeat;
      background-size: 75%;

      &::before {
        ${({ as }) => as === 'toggle' && css`
          background: ${theme.colors.success};
          border-color: ${theme.colors.success};
        `}
      }

      &::after {
        ${({ as }) => as === 'toggle' && css`
          left: auto;
          right: 0;
        `}
      }
    }

    &:checked ~ .a-radio {
      border: ${theme.borders[2]} #1785ff;

      &::after {
        content: '';
        top: 0;
        left: 0;
        right: 0;
        bottom: 0;
        width: 40%;
        height: 40%;
        margin: auto;
        position: absolute;
        border-radius: 30%;
        background: #1785ff;
      }
    }
  }

  ${({ readonly }) => readonly && css`
    opacity: 0.38;
    pointer-events: none;
  `}

  &[type='checkbox'],
  &[type='radio'] {
    label {
      align-items: center;
      cursor: pointer;
      display: flex;
    }
  }
`;

export const HelperText = styled.span`
  color: ${({ theme }) => theme.isDark
    ? theme.colors.whites[8] : theme.colors.blacks[8]};
  display: block;
  font-size: ${theme.fontSizes[0]}px;
  padding:
    ${theme.space[1] / 2}px
    ${theme.space[1] + theme.space[1] / 2}px
    0
    ${theme.space[1] + theme.space[1] / 2}px;
`;

export default Wrapper;
