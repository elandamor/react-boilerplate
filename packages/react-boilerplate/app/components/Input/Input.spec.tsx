// Input.spec.tsx
import * as React from 'react';
import * as renderer from 'react-test-renderer';
import 'jest-styled-components';

import Input from './index';

describe('<Input />', () => {
  it('should display a form input', () => {
    const component = renderer
      .create(
        <Input
          id="username"
          label="Username"
          name="username"
          onChange={() => null}
        />,
      )
      .toJSON();

    expect(component).toMatchSnapshot();
  });
});
