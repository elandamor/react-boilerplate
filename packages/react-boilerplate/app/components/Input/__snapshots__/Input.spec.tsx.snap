// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Input /> should display a form input 1`] = `
<div
  className="c-input-wrapper "
>
  <label
    htmlFor="username"
  >
    <span
      className="a-label"
    >
      Username
    </span>
    <input
      className="a-input"
      id="username"
      name="username"
      onChange={[Function]}
      type="text"
    />
  </label>
</div>
`;
